presto-main/src/main/java/com/facebook/presto/resourcemanager/ResourceManagerClusterStateProvider.java

```diff
@@ -0,0 +1,358 @@
+    {
+        List<MemoryInfo> memoryInfos = nodeStatuses.values().stream()
+                .map(nodeStatus -> nodeStatus.getNodeStatus().getMemoryInfo())
+                .collect(toImmutableList());
+
+        Map<MemoryPoolId, Long> counts = nodeQueryStates.values().stream()
+                .flatMap(state -> state.getActiveQueries().stream())
+                .collect(groupingBy(query -> query.getBasicQueryInfo().getMemoryPool(), counting()));
+        // Add defaults when queries are not running
+        counts.putIfAbsent(GENERAL_POOL, 0L);
+        if (isReservedPoolEnabled) {
+            counts.putIfAbsent(RESERVED_POOL, 0L);
+        }
+
+        Map<MemoryPoolId, ClusterMemoryPoolInfo> collect = counts.entrySet().stream().collect(toImmutableMap(Map.Entry::getKey, entry -> {
+            ClusterMemoryPool pool = new ClusterMemoryPool(entry.getKey());
+            pool.update(memoryInfos, toIntExact(entry.getValue()));
+            return pool.getClusterInfo();
+        }));
+        return collect;
+    }
+
+    public synchronized Map<String, MemoryInfo> getWorkerMemoryInfo()
+    {
+        return nodeStatuses.entrySet().stream().collect(toImmutableMap(e -> {
+            String nodeIdentifier = e.getValue().getNodeStatus().getNodeId();
+            String nodeHost = URI.create(e.getValue().getNodeStatus().getExternalAddress()).getHost();
+            return nodeIdentifier + " [" + nodeHost + "]";
+        }, e -> e.getValue().getNodeStatus().getMemoryInfo()));
+    }
+
+    private static class NodeQueryState
+    {
+        private final String nodeId;
+        private final ConcurrentMap<QueryId, Query> activeQueries = new ConcurrentHashMap<>();
+
+        @GuardedBy("this")
+        private final Map<QueryId, Query> completedQueries;
+
+        private final int maxCompletedQueries;
+        private final long queryExpirationTimeoutMillis;
+        private final long completedQueryExpirationTimeoutMillis;
+
+        public NodeQueryState(String nodeId, int maxCompletedQueries, long queryExpirationTimeoutMillis, long completedQueryExpirationTimeoutMillis)
+        {
+            this.nodeId = requireNonNull(nodeId, "nodeId is null");
+            completedQueries = new LinkedHashMap<>();
+            checkArgument(maxCompletedQueries > 0);
+            checkArgument(queryExpirationTimeoutMillis > 0);
+            checkArgument(completedQueryExpirationTimeoutMillis > 0);
+            this.maxCompletedQueries = maxCompletedQueries;
+            this.queryExpirationTimeoutMillis = queryExpirationTimeoutMillis;
+            this.completedQueryExpirationTimeoutMillis = completedQueryExpirationTimeoutMillis;
+        }
+
+        public synchronized NodeQueryState updateQuery(BasicQueryInfo basicQueryInfo)
+        {
+            if (activeQueries.computeIfPresent(basicQueryInfo.getQueryId(), (queryId, query) -> query.updateQueryInfo(basicQueryInfo)) == null) {
```

Comment:
If the query is already completed then why not just put it in completed queries list and remove from active queries here? We do have query completion state in the BasicQueryInfo.